fw/cache.c

```diff
@@ -2650,17 +2623,64 @@ tfw_cache_add(TfwHttpResp *resp, tfw_http_cache_cb_t action)
 	action((TfwHttpMsg *)resp);
 }
 
+/*
+ * Invalidate all cache entries that match the request. This is implemented by
+ * making the entries stale.
+ *
+ */
 static int
-tfw_cache_purge_immediate(TfwHttpReq *req)
+tfw_cache_purge_invalidate(TDB *db, TfwHttpReq *req)
+{
+	TdbIter iter;
+	TfwCacheEntry *ce = NULL;
+	unsigned long key = tfw_http_req_key_calc(req);
+
+	iter = tdb_rec_get(db, key);
+	if (TDB_ITER_BAD(iter))
+		return 0;
+
+	ce = (TfwCacheEntry *)iter.rec;
+	do {
+		if (tfw_cache_entry_key_eq(db, req, ce))
+			ce->lifetime = 0;
+		tdb_rec_next(db, &iter);
+		ce = (TfwCacheEntry *)iter.rec;
+	} while (ce);
+
+	return 0;
+}
+
+/*
+ * Remove all cache entries that match the request.
+ */
+static int
+tfw_cache_purge_immediate(TDB *db, TfwHttpReq *req)
 {
-	TDB *db = node_db();
 	unsigned long key = tfw_http_req_key_calc(req);
 
 	tdb_entry_remove(db, key, &tfw_cache_rec_eq_req, req, false);
 
 	return 0;
 }
 
+static int
+tfw_cache_purge_invoke_impl(TfwHttpReq *req,
+			    int purge_impl(TDB *db, TfwHttpReq *req))
+{
+	int ret;
+
+	if (cache_cfg.cache == TFW_CACHE_REPLICA) {
+		int nid;
+
+		for_each_node_with_cpus(nid)
+			ret = purge_impl(c_nodes[nid].db, req);
```

Comment:
Shouldn't we "remember" the first non-zero value? At the moment if a call fails, but the next one succeed, we return 0

Reply:
We could, but as I see there is no difference for caller. In situation where we have 4 nodes and one fails to purge, but three nodes purged successfully might be treated as success of as fail it depends on the rest code. But for now it's the same.